moveit_ros/planning_interface/test/plan_using_subframes.cpp

```diff
@@ -0,0 +1,299 @@
+  bool found_box_subframe = false;
+  for (size_t i = 0; i < box_subframe_names.size() && !found_box_subframe; ++i)
+  {
+    if (box_subframe_names[i] == subframe_name)
+    {
+      found_box_subframe = true;
+      box_subframe_msg = box.subframe_poses[i];
+      break;
+    }
+  }
+  ASSERT_TRUE(found_box_subframe);
+
+  poseMsgToEigen(box_subframe_msg, box_subframe_in_world);
+
+  ASSERT_LT(std::abs(box_subframe_in_world.translation()[0] - tip_in_world.translation()[0]), 0.1 + 1e-5);
+  ASSERT_LT(std::abs(box_subframe_in_world.translation()[1] - tip_in_world.translation()[1]), 0.1 + 1e-5);
+  ASSERT_LT(std::abs(box_subframe_in_world.translation()[1] - tip_in_world.translation()[1]), 0.1 + 1e-5);
+}
+
+TEST(TestPlanUsingSubframes, SubframesTests)
+{
+  ros::NodeHandle nh;
+  ros::AsyncSpinner spinner(1);
+  spinner.start();
+
+  geometry_msgs::Pose tip_in_hand_msg, hand_in_world_msg, tip_in_world_msg;
+  Eigen::Affine3d tip_in_hand, tip_in_world, hand_in_word;
+
+  ros::WallDuration(1.0).sleep();
+  moveit::planning_interface::PlanningSceneInterface planning_scene_interface;
+  moveit::planning_interface::MoveGroupInterface group("panda_arm");
+  group.setPlanningTime(10.0);
+
+  spawnCollisionObjects(planning_scene_interface);
+  moveit_msgs::AttachedCollisionObject att_coll_object;
+  att_coll_object.object.id = "cylinder";
+  att_coll_object.link_name = "panda_hand";
+  att_coll_object.object.operation = att_coll_object.object.ADD;
+  ROS_INFO_STREAM("Attaching cylinder to robot.");
+  planning_scene_interface.applyAttachedCollisionObject(att_coll_object);
+
+  // Define a pose in the robot base.
+  tf2::Quaternion orientation, orientation_1, target_orientation;
+  geometry_msgs::PoseStamped fixed_pose, temp_pose_stamped;
+  fixed_pose.header.frame_id = "panda_link0";
+  fixed_pose.pose.position.y = -.4;
+  fixed_pose.pose.position.z = .3;
+  orientation.setRPY(0, (-20.0 / 180.0 * M_PI), 0);
+  fixed_pose.pose.orientation = tf2::toMsg(orientation);
+
+  // We multiply tf2 quaternions to chain together a few rotations. This is usually simpler and easier
+  // to debug than figuring out a series of euler angles in your head.
+  tf2::Quaternion flip_around_y;  // This is used to rotate the orientation of the target pose.
+  flip_around_y.setRPY(0, (180.0 / 180.0 * M_PI), 0);
+
+  ROS_INFO_STREAM("Moving to bottom of box with cylinder tip");
+  temp_pose_stamped.header.frame_id = "box/bottom";
+
+  // When constructing the target pose for the robot, we multiply the quaternions to get the
+  // target orientation and convert the result to a ``geometry_msgs/orientation`` message.
+  orientation_1.setRPY(-(90.0 / 180.0 * M_PI), 0, 0);
+  target_orientation = flip_around_y * orientation_1;
+  temp_pose_stamped.pose.orientation = tf2::toMsg(target_orientation);
+  temp_pose_stamped.pose.position.z = 0.01;
+  moveToCartPose(temp_pose_stamped, group, "cylinder/tip");
+  testNearSubframe("bottom", &planning_scene_interface, &group);
```

Comment:
When using tests in a sub-function, use SCOPED_TRACE so you can easily track down where the error is coming from:

```
  {
    SCOPED_TRACE("Initial orientation");
    ....
    testNearSubframe(...);
  }
```